File: robot/ros_ws/src/hrc_utils/test/test_utils.cpp
#include "hrc_utils/utils.hpp"
#include "hrc_utils/testing_utils.hpp"

#include <gtest/gtest.h>

// Test the robotToMap function
TEST(TestUtils, RobotToMap)
{
    geometry_msgs::msg::Pose2D robot_pose;
    geometry_msgs::msg::Point32 robot_point;
    geometry_msgs::msg::Point32 map_point;

    // Check robot at the origin
    robot_pose.x = 0.0;
    robot_pose.y = 0.0;
    robot_pose.theta = 0.0;

    robot_point.x = 1.0;
    robot_point.y = 1.0;

    hrc_utils::robotToMap(robot_pose, robot_point, map_point);
    ASSERT_NEAR(map_point.x, 1.0, HRC_UTILS__TESTING_FLOAT_ASSERTION_PRECISION);
    ASSERT_NEAR(map_point.y, 1.0, HRC_UTILS__TESTING_FLOAT_ASSERTION_PRECISION);

    // Check robot still at the origin but rotated 90 degrees
    robot_pose.theta = M_PI_2;

    hrc_utils::robotToMap(robot_pose, robot_point, map_point);
    ASSERT_NEAR(map_point.x, -1.0, HRC_UTILS__TESTING_FLOAT_ASSERTION_PRECISION);
    ASSERT_NEAR(map_point.y, 1.0, HRC_UTILS__TESTING_FLOAT_ASSERTION_PRECISION);

    // Check robot not at the origin and still at 90 degrees
    robot_pose.x = 0.5;
    robot_pose.y = 0.5;

    hrc_utils::robotToMap(robot_pose, robot_point, map_point);
    ASSERT_NEAR(map_point.x, -0.5, HRC_UTILS__TESTING_FLOAT_ASSERTION_PRECISION);
    ASSERT_NEAR(map_point.y, 1.5, HRC_UTILS__TESTING_FLOAT_ASSERTION_PRECISION);
}

TEST(TestUtils, FloatComparison)
{
    // Double
    ASSERT_TRUE(hrc_utils::floatEqual(1.0, 1.0, HRC_UTILS__TESTING_FLOAT_ASSERTION_PRECISION));
    ASSERT_FALSE(hrc_utils::floatEqual(1.0, 1.1, HRC_UTILS__TESTING_FLOAT_ASSERTION_PRECISION));

    // Float
    ASSERT_TRUE(hrc_utils::floatEqual(1.0f, 1.0f, (float) HRC_UTILS__TESTING_FLOAT_ASSERTION_PRECISION));
    ASSERT_FALSE(hrc_utils::floatEqual(1.0f, 1.1f, (float) HRC_UTILS__TESTING_FLOAT_ASSERTION_PRECISION));
}

File: robot/ros_ws/src/hrc_utils/test/test_utils.py
import math
import os

import geometry_msgs.msg as geo_msgs
import hrc_utils
import pytest


def test_robot_to_map():
    """Test the robotToMap function."""
    # Check robot at the origin
    robot_pose = geo_msgs.Pose2D(x=0.0, y=0.0, theta=0.0)
    robot_point = geo_msgs.Point32(x=1.0, y=1.0)
    map_point = hrc_utils.robot_to_map(robot_pose, robot_point)

    assert map_point.x == pytest.approx(1.0)
    assert map_point.y == pytest.approx(1.0)

    # Check robot still at the origin but rotated 90 degrees
    robot_pose.theta = math.pi / 2
    map_point = hrc_utils.robot_to_map(robot_pose, robot_point)

    assert map_point.x == pytest.approx(-1.0)
    assert map_point.y == pytest.approx(1.0)

    # Check robot not at the origin and still at 90 degrees
    robot_pose.x = 0.5
    robot_pose.y = 0.5
    map_point = hrc_utils.robot_to_map(robot_pose, robot_point)

    assert map_point.x == pytest.approx(-0.5)
    assert map_point.y == pytest.approx(1.5)


def test_get_herminebot_model():
    """Test the get_herminebot_model function."""
    assert hrc_utils.get_herminebot_model() == 'diff'

    os.environ['HERMINEBOT_MODEL'] = 'omni'
    assert hrc_utils.get_herminebot_model() == 'omni'


if __name__ == '__main__':
    pytest.main(['-v'])
